src/util/retry.rs

```diff
@@ -0,0 +1,253 @@
+	}
+}
+
+/// Represents the context of a retry operation.
+///
+/// The context holds key information about the retry operation
+/// such as how many attempts have been made till now, the accumulated
+/// delay between retries, and the error which triggered the retry.
+pub struct RetryContext<'a, E: Error> {
+	/// The number attempts made until now, before attempting next retry.
+	attempts_made: u32,
+
+	/// The amount of artificial delay we have already waited in between previous
+	/// attempts. Does not include the time taken to execute the operation that is being retried.
+	accumulated_delay: Duration,
+
+	/// The error encountered in previous attempt.
+	error: &'a E,
+}
+
+/// The exponential backoff strategy is a retry approach that doubles the delay between retries.
+/// A combined exponential backoff and jitter strategy is recommended that is ["Exponential Backoff and Jitter"](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/).
+/// This is helpful to avoid [Thundering Herd Problem](https://en.wikipedia.org/wiki/Thundering_herd_problem).
+pub struct ExponentialBackoffRetryPolicy {
+	/// The base delay duration for the backoff algorithm. First retry is `base_delay` after first attempt.
+	pub base_delay: Duration,
+}
+
+impl<E: Error> RetryPolicy<E> for ExponentialBackoffRetryPolicy {
+	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration> {
+		let backoff_factor = 2_u32.pow(context.attempts_made) - 1;
+		let delay = self.base_delay * backoff_factor;
+		Some(delay)
+	}
+}
+
+/// Decorates the given `RetryPolicy` and adds jitter (random delay) to it. This can make retries
+/// more spread out and less likely to all fail at once.
+pub struct JitteredRetryPolicy<T: RetryPolicy<E>, E: Error> {
+	/// The underlying retry policy to use.
+	inner_policy: T,
+	/// The maximum amount of random jitter to apply to the delay.
+	max_jitter: Duration,
+	phantom: PhantomData<E>,
+}
+
+impl<T: RetryPolicy<E>, E: Error> RetryPolicy<E> for JitteredRetryPolicy<T, E> {
+	fn next_delay(&self, context: &RetryContext<E>) -> Option<Duration> {
+		if let Some(base_delay) = self.inner_policy.next_delay(context) {
+			let mut rng = rand::thread_rng();
+			let jitter = Duration::from_micros(rng.gen_range(0..self.max_jitter.as_micros() as u64));
+			Some(base_delay + jitter)
+		} else {
+			None
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use std::sync::atomic::{AtomicU32, Ordering};
+	use std::sync::Arc;
+
+	use crate::error::VssError;
+
+	use super::*;
+
+	#[tokio::test]
+	async fn test_async_retry_success() {
+		let base_delay = Duration::from_millis(50);
+		let max_attempts = 3;
+		let max_total_delay = Duration::from_secs(6000);
+		let max_jitter = Duration::from_millis(10);
+
+		let exponential_backoff_jitter_policy = ExponentialBackoffRetryPolicy { base_delay }
+			.skip_retry_on_error(|e| matches!(e, VssError::InvalidRequestError(..)))
+			.with_max_attempts(max_attempts)
+			.with_max_total_delay(max_total_delay)
```

Comment:
Could you have the closure return an error based on the number of attempts so we can exercise this logic?